File: android/src/com/ollix/moui/OpenGLView.java
// Copyright (c) 2014 Ollix. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ---
// Author: olliwang@ollix.com (Olli Wang)

package com.ollix.moui;

import com.ollix.moui.OpenGLRenderer;
import android.content.Context;
import android.graphics.PixelFormat;
import android.opengl.GLSurfaceView;
import android.view.MotionEvent;
import android.util.Log;

public class OpenGLView extends GLSurfaceView {

  private boolean mHandlingEvent;
  private long mMouiViewPointer;
  private native void handleEvent(long mouiViewPointer, int action, float x,
                                  float y);
  private native boolean shouldHandleEvent(long mouiViewPointer, float x,
                                           float y);

  public OpenGLView(Context context, long mouiViewPointer) {
    super(context);
    mHandlingEvent = false;
    mMouiViewPointer = mouiViewPointer;

    setEGLContextClientVersion(2);  // adopts OpenGL ES 2.0 context
    setZOrderOnTop(true);
    setEGLConfigChooser(8,  // red size
                        8,  // green size
                        8,  // blue size
                        8,  // alpha size
                        24,  // enables depth
                        8);  // enables stencil
    getHolder().setFormat(PixelFormat.TRANSLUCENT);

    setRenderer(new OpenGLRenderer(mouiViewPointer));
    setRenderMode(GLSurfaceView.RENDERMODE_WHEN_DIRTY);
  }

  // Returns the value returned by native view's HandleEvent() method.
  // If false, the event will be passed to the next responser.
  @Override
  public boolean onTouchEvent(MotionEvent event) {
    // Do nothing if the event is not handling and is not the initial action.
    if (!mHandlingEvent && event.getAction() != MotionEvent.ACTION_DOWN) {
      return false;
    }
    // Determines the action location.
    int[] coords = new int[2];
    getLocationInWindow(coords);
    final float x = event.getRawX() - coords[0];
    final float y = event.getRawY() - coords[1];
    // If it's an initial action, checks if it should handle consequent events.
    if (!mHandlingEvent && !shouldHandleEvent(mMouiViewPointer, x, y)) {
      return false;
    }
    mHandlingEvent = true;
    // Handles the event in corresponded moui view.
    handleEvent(mMouiViewPointer, event.getAction(), x, y);
    // Resets mHandlingEvent if the action is complete.
    if (event.getAction() == MotionEvent.ACTION_UP ||
        event.getAction() == MotionEvent.ACTION_CANCEL) {
      mHandlingEvent = false;
    }
    return true;  // Returns true means this event has consumed.
  }
}

File: android/src/main/java/com/ollix/moui/OpenGLView.kt
/*
 * Copyright (c) 2017 Ollix. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ---
 * Author: olliwang@ollix.com (Olli Wang)
 */

package com.ollix.moui

import android.content.Context
import android.opengl.EGL14
import android.opengl.EGLConfig
import android.opengl.EGLContext
import android.opengl.EGLDisplay
import android.opengl.EGLSurface
import com.ollix.moui.View

class OpenGLView(context: Context, mouiViewPtr: Long)
        : View(context, mouiViewPtr) {
    private var eglDisplay: EGLDisplay = EGL14.EGL_NO_DISPLAY
    private var eglContext: EGLContext = EGL14.EGL_NO_CONTEXT
    private var eglSurface: EGLSurface = EGL14.EGL_NO_SURFACE

    override fun createDrawable(): Boolean {
        EGL14.eglGetCurrentContext()
        eglDisplay = EGL14.eglGetDisplay(EGL14.EGL_DEFAULT_DISPLAY)
        if (eglDisplay == EGL14.EGL_NO_DISPLAY) {
            return false
        }

        val version = IntArray(2)
        if (!EGL14.eglInitialize(eglDisplay, version, 0, version, 1)) {
            return false
        }

        /* Initializes the config. */
        var configAttributes = intArrayOf(
            EGL14.EGL_RED_SIZE, 8,
            EGL14.EGL_GREEN_SIZE, 8,
            EGL14.EGL_BLUE_SIZE, 8,
            EGL14.EGL_ALPHA_SIZE, 8,
            EGL14.EGL_STENCIL_SIZE, 8,
            EGL14.EGL_RENDERABLE_TYPE, EGL14.EGL_OPENGL_ES2_BIT,
            EGL14.EGL_NONE
        )
        val configs: Array<EGLConfig?> = arrayOfNulls<EGLConfig>(1)
        val numOfConfigs = IntArray(1)
        EGL14.eglChooseConfig(eglDisplay, configAttributes, 0, configs, 0,
                              configs.size, numOfConfigs, 0)
        if (EGL14.eglGetError() != EGL14.EGL_SUCCESS) {
            return false
        }

        /** Initializes the context. */
        var contextAttributes = intArrayOf(
            EGL14.EGL_CONTEXT_CLIENT_VERSION, 2,
            EGL14.EGL_NONE
        )
        eglContext = EGL14.eglCreateContext(eglDisplay, configs[0], eglContext,
                                            contextAttributes, 0)
        if (EGL14.eglGetError() != EGL14.EGL_SUCCESS) {
            return false
        }

        /** Initializes the surface. */
        var surfaceAttributes = intArrayOf(
            EGL14.EGL_NONE
        )
        eglSurface = EGL14.eglCreateWindowSurface(eglDisplay, configs[0],
                                                  holder.surface,
                                                  surfaceAttributes, 0)
        if (EGL14.eglGetError() != EGL14.EGL_SUCCESS) {
            return false
        }

        return true
    }

    override fun destroyDrawable() {
        if (eglDisplay != EGL14.EGL_NO_DISPLAY) {
            EGL14.eglMakeCurrent(eglDisplay,
                                 EGL14.EGL_NO_SURFACE,
                                 EGL14.EGL_NO_SURFACE,
                                 EGL14.EGL_NO_CONTEXT)
            EGL14.eglDestroyContext(eglDisplay, eglContext)
            EGL14.eglReleaseThread()
            EGL14.eglTerminate(eglDisplay)
        }
        eglDisplay = EGL14.EGL_NO_DISPLAY
        eglContext = EGL14.EGL_NO_CONTEXT
        eglSurface = EGL14.EGL_NO_SURFACE
    }

    override fun prepareDrawable() {
        EGL14.eglMakeCurrent(eglDisplay, eglSurface, eglSurface, eglContext)
    }

    override fun presentDrawable() {
        EGL14.eglSwapBuffers(eglDisplay, eglSurface)
    }
}
